src/backend/distributed/master/master_create_shards.c

```diff
@@ -409,11 +456,20 @@ CreateReferenceTableShard(Oid distributedTableId)
 	 */
 	LockShardDistributionMetadata(shardId, ExclusiveLock);
 
-	CreateShardPlacements(distributedTableId, shardId, ddlCommandList, relationOwner,
-						  workerNodeList, workerStartIndex, replicationFactor);
-
 	InsertShardRow(distributedTableId, shardId, shardStorageType, shardMinValue,
 				   shardMaxValue);
+
+	connectionList = CreateShardPlacements(distributedTableId, shardId, ddlCommandList,
+										   relationOwner, workerNodeList,
+										   workerStartIndex,
+										   replicationFactor, emptyTable,
+										   createInSeperateTransaction);
+
+	foreach(connectionCell, connectionList)
+	{
+		MultiConnection *connection = (MultiConnection *) lfirst(connectionCell);
+		UnclaimConnection(connection);
```

Comment:
For readability, consider structuring the code as something like:
```
void
CreateShardsOnWorkers(List *newShardList, ..., bool useExclusiveConnections, bool createInSubtransaction)
{
    List *claimedConnectionList = NIL;
    int connectionFlags = 0;

    if (createInSubtransaction)
    {
        connectionFlags = FORCE_NEW_CONNECTION;
    }

    foreach(newShardCell, newShardList)
    {
        ...
        for (...)
        {
            ...
            connection = GetPlacementConnection(connectionFlags, ...);
            if (useExclusiveConnections)
            {
                /* use separate connection per placement to allow subsequent COPY */
                ClaimConnectionExlusively(connection);
                claimedConnectionsList = lappend(claimedConnectionsList, connection);
            }
            if (!createInSubtransaction)
            {
                RemoteTransactionBeginIfNecessary(connection);
            }
            ...
            WorkerCreateShardPlacement(connection, ...);
            ...
        }
    }

    foreach(connectionCell, claimedConnectionList)
    {
        ...
        UnclaimConnection(connection);
    }
}
```

That would make the overall procedure much easier to follow, since all the connection handling logic is in one place.